File: JS/cedrus/src/models/ClassificationModel.ts
import { isArray } from "@juniper-lib/util";
import { ClassificationCaveatModel, isClassificationCaveatModel } from "./ClassificationCaveatModel";
import { ClassificationLevelModel, isClassificationLevelModel } from "./ClassificationLevelModel";
import { IDOrName } from "./IDOrName";
import { INamed, isINamed } from "./INamed";
import { ITypeStamped, isTypeStamped } from "./ITypeStamped";

export interface ClassificationModel extends INamed, ITypeStamped<"classification"> {
    description: string;
    level: ClassificationLevelModel;
    caveats: ClassificationCaveatModel[];
}

export function isClassificationModel(obj: unknown): obj is ClassificationModel {
    return isTypeStamped("classification", obj)
        && isINamed(obj)
        && "level" in obj
        && isClassificationLevelModel(obj.level)
        && "caveats" in obj
        && isArray(obj.caveats)
        && obj.caveats.every(isClassificationCaveatModel);
}

export interface SetClassificationInput {
    level: IDOrName;
    caveats: IDOrName[];
}

File: NETCore/Cedrus/Controllers/V1/Models/ClassificationModel.cs
using Juniper.Cedrus.Entities;

namespace Juniper.Cedrus.Controllers.V1;

public class ClassificationModel
{
    public string TypeStamp => "classification";
    public int? Id { get; }
    public string? Name { get; }
    public string? Description { get; }
    public string? CaveatString { get; }

    public ClassificationLevelModel Level { get; }

    public ClassificationCaveatModel[] Caveats { get; }

    public ClassificationModel(Classification? classification)
    {
        Id = classification?.Id;
        Name = classification?.Name;
        Description = classification?.Description;
        CaveatString = classification?.CaveatsString;
        Level = new ClassificationLevelModel(classification?.Level);
        Caveats = classification
            ?.Caveats
            ?.Select(c => new ClassificationCaveatModel(c))
            ?.ToArray()
            ?? Array.Empty<ClassificationCaveatModel>();
    }
}
